### DISCUSSION TITLE
OpenDTU-onBattery schematic, BOM, and PCB for 2 Victrons MPPTs

### DISCUSSION THREAD

**Discussion by helgeerbe** [hoylabs/OpenDTU-OnBattery#138] [Show and tell]
Actual OpenDTU-onBattery supports only one Victron MPPT, but I plan to extend it.  So I created a schema for 2 MPPTs

I power my OpenDTU-onBattery with a 5V DIN rail power supply. Therefore I use a voltage regulator to get the 3.3V. USB is only used for debug purpose.

## Schematic
![OpenDTU](https://user-images.githubusercontent.com/59169507/225047048-251504e9-d62a-49d9-9364-d83cb7369ad9.svg)

## Bill of material

| Ref | Qnty | Value  | Footprint  | Description   | 
|---|---|---|---|---|
| C1  |  1 | 1000UF / 16 V  | Capacitor_THT:C_Radial_D10.0mm_H16.0mm_P5.00mm  | Polarized capacitor |
| C2 |  1 |  100nF 50V |  Capacitor_THT:C_Disc_D6.0mm_W2.5mm_P5.00mm | Unpolarized capacitor |
| J1  |  1 | Victron 1  | TerminalBlock:TerminalBlock_Altech_AK300-3_P5.00mm | Generic screw terminal, single row, 01x03 |
| J2  |  1 | Victron 2  | TerminalBlock:TerminalBlock_Altech_AK300-3_P5.00mm | Generic screw terminal, single row, 01x03 |
| J3  |  1 | PWR 5V | TerminalBlock:TerminalBlock_Altech_AK300-2_P5.00mm | Generic screw terminal, single row, 01x02 |
| U1  |  1 | Wemos esp32 D1 mini  | ESP32_mini:ESP32_mini  |  |
| U2  |  1 | MCP1700x-330xxTO  | Package_TO_SOT_THT:TO-92_Inline  | 250mA Low Quiscent Current LDO, 3.3V output, TO-92 |
| U3  |  1 |  Level Shifter 3.3V to 5V |   4bit level shifter| |
| U4  |  1 | NRF24L01_Breakout+  |  RF_Module:nRF24L01+_Breakout | Ultra low power 2.4GHz RF Transceiver, Carrier PCB |

### PCB

Still waiting for the PCB. I hope I did the wiring correctly.

![OpenDTU-onBattery_PCB](https://user-images.githubusercontent.com/59169507/225053489-6f8f0f6e-ab55-432e-9cd9-53c3ec82187a.png)

**Comment by helgeerbe** [comment]
![pcb](https://user-images.githubusercontent.com/59169507/227919565-533e0871-6b26-4988-b9d9-065b844a171c.png)

Recieved my pcp and it works. Just one copy/paste error in the labeling of the serial interfaces. Prepared for one additional Victron MPPT charger.

**Reply by berni2288** [reply]
looks really epic!

**Reply by helgeerbe** [reply]
yes, my first pcb. I use FreeCad quite often. But never used KiCad. Both are quite a good team.

**Comment by swingstate** [comment]
That is really cool! Designing and building a workable PCB is something I still have on my todo list for this life :)

My 2x2 panels are facing SSW / SSO so this looks like I could keep them as they are and use two MPPT chargers. Amazing work, happy testing! :)

**Comment by Manos1966** [comment]
This is amazing! How can you have a PCB made and how expensive is it?
Can I have/order/pay one such PCB of yours?

**Reply by helgeerbe** [reply]
@Manos1966 I can share the kiCad files and if you like you can order it online. I paid around 13€ for 3 PCBs (minim quantity) at https://aisler.net

So I have 2 spare parts. If you want one, I can send one to you at cost (4 € + postage fee). This is only the plain PCB, so you have to buy the other parts and solder it on the PCB. And maybe you should print yourself an housing, if you have a 3d printer. I can share also a simple model for that.

![ezgif-2-6d923336bf](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/59169507/640149a4-0ae0-4572-ba16-fbbbde9e0ddc)

**Reply by Manos1966** [reply]
Hi @helgeerbe , **that** would be great! 
PayPal? send me a private message with total cost (I am in Germany).
Also for the 3D model please. I am only used to printing PLE, I think I would need an ABS print for that but I guess I could even find a company that can print it ABS for me.
I am using the bigger ESP32s for the Test configurations because the PINs are already soldered but I can get the smaller Wemos esp32 D1 mini again to try.
I am amazed to learn that a company in Germany can print PCBs, starting at 13EUR 👍 👍 👍

**Reply by helgeerbe** [reply]
@Manos1966 will send you an email. Would be next week. I'm away this weekend.

**Reply by Manos1966** [reply]
Many Thanks! Enjoy the long Weekend!

**Reply by Manos1966** [reply]
Hi @helgeerbe , 
don't forget to send me the email with the total cost (incl postage)...
I have a question regarding aslaier.net, what 'Surface Finish' did you order for your pcb? (Surface Finish: HASL, or  ENIG)?

**Reply by helgeerbe** [reply]
Was HASL. Sorry, I forgot it totally. Do you have an email address for me?

**Reply by helgeerbe** [reply]
Kompaktbrief up to 50g should be fine. So in total it's 5€

**Reply by Manos1966** [reply]
It arrived! 
Many thanks Helge :-)

**Reply by helgeerbe** [reply]
Have fun :)

**Reply by Manos1966** [reply]
@helgeerbe By the way, please post also the 3D Model to put the pcb in ;-)

**Reply by helgeerbe** [reply]
[OpenDTU-onBattery-2.zip](https://github.com/helgeerbe/OpenDTU-OnBattery/files/11633236/OpenDTU-onBattery-2.zip)
I also included the original freeCad model, if you like to modify the case. To mount the PCB you need M3 screws and HEAT-SET inserts (Einpressmuttern).

**Reply by Manos1966** [reply]
@helgeerbe Almost finished soldering, will not earn any prize for my soldering skills but, I will be happy if it works.
Question: Can I power the system from the USB port of the ESP32 or, must I use the J3 connector?

**Reply by Manos1966** [reply]
Many thanks Helge, it seems I managed to solder everything and it is (surprisingly) working!
![20230623_023938](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/129660360/64d23733-6fdc-48ef-a428-cf52bad7389e)

**Reply by Manos1966** [reply]
By the way, a probably stupid question: 
Why do you put both a
C1 | 1 | 1000UF / 16 V |
and a 
C2 | 1 | 100nF 50V ?

**Reply by helgeerbe** [reply]
Der Kerko ist dem Elko parallelgeschaltet.
Das hat den Hintergrund,das ein Elko zum "filtern" von kurzen Spannungsspitzen und HF viel zu träge ist. 

Schlechte Netzteile würden sonst ggf. Probleme machen.

**Reply by Manos1966** [reply]
Helge, just asking as I struggle to learn how to modify things with kiCAD: 
You do not have by any chance a pcb with an ESP32 DevKit C V4 (instead of the D1 Mini) already designed, have you?

**Reply by helgeerbe** [reply]
I don't but try this:
- https://www.snapeda.com/parts/ESP32-DEVKITC-32D/Espressif%20Systems/view-part/
- https://grabcad.com/library/esp32-devkitc-v4-1

**Comment by madmartin** [comment]
@helgeerbe 

1st question: why did you use a voltage regulator on the pcb, connected to the +5V terminal? There is a more powerful 5V->3,3V Regulator on the ESP32-NodeMCU PCB, and you can feed +5V into the NodeMCU through one of the pins (when you dont want to have the microUSB as the power source).
Was there a specific reason for this?

2nd: if you want to share the kicad files, just create another github project with it..... I would propose a structure with one directory per PCB variant - I think I will create also one, with CANbus connection, which I would offer to add to such a repo.

**Reply by helgeerbe** [reply]
1. I'm using a DIN rail power supply, which seem to be not so good and the esp runs a little bit unreliable. I'm using this voltage regulator also for my battery IOT project. The capcitor supports the nrf and the esp. Since I use this setup, I don't have any issues regarding to a bad power supply any more. 
2. Agree, I coud set up another repo for that.

**Reply by swingstate** [reply]
are you sharing the PCB file somewhere? I am going to add 2 100/30 MPPTs to my setup and was thinking to start using the VE.Direct.

**Reply by helgeerbe** [reply]
[OpenDTU-kiCad.zip](https://github.com/helgeerbe/OpenDTU-OnBattery/files/11610454/OpenDTU-kiCad.zip)

But have in mind, that actually only one battery charger is supported. I just added the second connection for future enhancements.

**Reply by swingstate** [reply]
Ah, good to know. Can you tell if 2 chargers would cause any issue, if only one is connected to the ESP?

**Reply by helgeerbe** [reply]
No, that is not a problem. But you shouldn't use gpio 13 and 26, because I saved them for the second charger.

**Comment by kloppy1984** [comment]
![20230601_205712](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/63321205/3faa3429-8bda-4687-bc07-027284c66cca)
Hier ein kurzes Bild meiner Platine.
Angeschlossen an einen Victron Laderegler, die Galvanische Trennung ist bei mir im Kabel zum Victron integriert.
Canbus ist vorbereitet, muss nur noch verdrahtet werden.

**Reply by madmartin** [reply]
Galvanische Trennung im Kabel? Woraus besteht die und warum soll die notwendig sein?

**Reply by Manos1966** [reply]
Wow! Ist das RJ45 als Stecker fuer das CANbus unten rechts?

@helgeerbe a propos CANbus, ist jede Batterie mit CANbus kompatibel? (erwaehnt wird nur PYLONTECH)

**Reply by kloppy1984** [reply]
@madmartin 
Ein optocopler ADUM1201 sorgt dafür das beide Stromktreise von einander getrennt sind. 

Siehe https://www.akkudoktor.net/forum/stell-dein-batterie-powerwall-projekt-vor/warnung-vor-selbstgebauten-ve-direct-usb-kabeln/

@Manos1966 
Richrig, unten ist der RJ 45 für den Canbus für die Pylontech Akkus... Da ich noch keine Belegung des Steckers gefunden habe, habe ich mir alle 8 Pins raus gelegt, mein Kumpel hat 4 Pylontech Akkus, muss das mal testen

Am esp32 ist zudem noch eine Antenne fürs beste WLAN festgelötet

**Reply by Manos1966** [reply]
@kloppy1984 "Ein optocopler ADUM1201" Du sprichst ueber dies, richtig? (Ausschnitt von OpenDTUonBattery Schematic)
![Zoom Optokopler](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/129660360/bb452657-8641-4948-b9b1-cd71b5aaf7ec)
Ich habe eine Frage diesbezueglich: 
Ich habe zur Zeit nur die Tx/Rx Kabel vom ESP32 *direkt* mit dem VICTRON angeschlossen und es funktioniert (ich habe ein Post von @helgeerbe gelesen, dass _kurzfristig_ sollte der ESP32 es aushalten). Eine Version mit optocopler loete ich gerade, sie ist aber noch nicht im Einsatz... 
Im Bild ist aber weder LV (3,3V von ESP32), noch HV (5V), noch GND vom VICTRON angeschlossen.
Frage1: Wenn ich das relay fuer das HUAWEI Netzteil oben rechts (ja, ich habe eins davon besorgt) loette, brauche ich nur Tx/Rx von VICTRON, richtig?
Frage2: Ohne das relay oben rechts, muss ich LV, GND von VICTRON sowie LV von der ESP32 anschliessen?

Sorry falls meine Fragen sehr dumm klingen, ich bin wie die Jungfrau zum Kind gekommen...
Ich koennte auch ein separates Post diesbezueglich oeffnen um nicht von PCB-Thema hier entgleiten.

**Reply by kloppy1984** [reply]
Das Rot eingerahmtes ist ein Levelshifter, er macht aus 5V Signale 3,3V Signale oder umgekehrt.
Dieser macht keine Trennung der Stromkreise.

<img width="158" alt="image" src="https://github.com/helgeerbe/OpenDTU-OnBattery/assets/63321205/7d8a4c19-7ed6-4b1b-aab8-ad0439a0c4f6">

Vorteil vom ADUM1201 ist, man ihn gut ins Kabel mit Schrupfschlauch verpacken.

<img width="800" alt="image" src="https://github.com/helgeerbe/OpenDTU-OnBattery/assets/63321205/2f4eef60-d99a-4837-894b-8e09dd27e622">

Zu deiner Frage: 
Dieser Chip LLC 3.3-5 (Roter kreis) macht aus 3,3V Signal ein 5V Signal.
Normalerweise kommt an LV 3,3V, an HV5,0V.
Dann weiß der Chip auf welche Voltlevel er das Signal weiterleiten soll.

Die Zeichnung ist NICHT VOLLSTÄNDIG!
Die Signal komme an LV4/ LV3 / LV2 / LV1 an und werden auf HV4 / HV3 / HV2 / HV1 mit 5V Level weiter geschickt.
An LV4 das Victron TX 3,3V vom ESP an und wird an HV4 zum Victron weiter geleteitet.
An LV3 das Victron TX 3,3V vom ESP an und wird an HV3 zum Victron weiter geleteitet.

- Die Masse vom Victron an den Chip fehlt, kann es zu Problem kommen!, da kein gemeinsames Potenzial existieren!
- Es fehlt 3,3V an LV. 
- Es fehlt  LV2 als Input für das Relaiy, welches das Relay schalten soll.

Antwort zu Frage 1:
Welche Funktion soll das Relais haben?
Aktuell wird es auf der Zeichnung nicht angesteuert, da LV2 fehlt. Es hängt nur an den Chip, ohne Funktion.
Für das Relaiy braucht es kein RX oder TX.

Antwort zu Frage 2:
Beim anschließen an den Victron benutze besser einen ADUM 1201 (Digitaler Isolator).
Wenn du das über LLC machen willst, sieht das anschließen wie folgt aus:

LV -> 3,3V vom ESP
GnD (neben LV) -> GND vom ESP
TX / RX von ESP an LV4 / LV3

Victron wird dann angeschlossen an:
HV -> 5V vom Victron
GnD (neben HV) -> GnD vom Victron
TX / RX von Victron an HV4/HV3

**Reply by Manos1966** [reply]
Vielen Dank @kloppy1984 !

wegen 

> TX / RX von ESP an LV4 / LV3
TX / RX von Victron an HV4/HV3

Ich glaube du meinst
TX von ESP geht an RX von Victron, sowie
RX von ESP geht an TX von Victron

**Comment by madmartin** [comment]
@kloppy1984 Ich habe mal zum ADUM1201 recherchiert... Grundsätzlich bezeifle ich daß wenn man die Anschlußhinweise von Victron befolgt es zur Zerstörung des SmartSolar kommen kann. Zu diesen Hinweisen zählt z.B. daß alle Komponenten sauber geerdet/auf Potentialausgleich gelegt werden sollen (an den dafür vorgesehenen Stellen!), aber weder das Batterie-Plus oder Minus noch die PV-Leitungen geerdet noch mit GND der Schaltungen verbunden werden sollen. 

Wenn Du aber einen ADUM1201 Isolierer verwenden willst, dann kann der gleichzeitig die Pegelwandlung von 3,3V nach 5V machen, man kann sich dann also ein Bauteil sparen. Kleiner Nachteil ggü. den käuflichen Level-Shifter-Platinchen ist der deutlich höhere Preis.
Interessant finde ich daß das KEIN Optokopler ist, diese Art Isolierung in einem IC ist mir neu (habe Elektroniker vor 30 Jahren gelernt.... da gab es vieles noch nicht was heute nur Centbeträge kostet).
Technische Details für interessiere gibts hier: https://www.reichelt.de/digitaler-isolator-2-kanal-2-7--5-v-1-mbps-150-ns-so-8-adum-1201-ar-p107978.html https://www.analog.com/media/en/technical-documentation/data-sheets/ADuM1200_1201.pdf

Fazit: Ich  bedanke mich  für die Horizont-Erweiterung!

**Comment by dirk-schulz** [comment]
Hello, absolutely great to get two Victron connected. Many thanks for the Schema.
But how to make the Config in OpenDTU, that it reads the 2nd Victron?
Can it be done via pin_mapping.json and If yes, how does it looks like?

**Reply by helgeerbe** [reply]
Sorry, at the moment only one victron is supported. The second one is for future changes.

**Reply by Manos1966** [reply]
Hi @helgeerbe 
If I understood correctly, many VICTRONS can be connected with each other.
Will OPENDTUonBattery in this case be able to see the sum of all VICTRONS or only one?

**Reply by schlimmchen** [reply]
Very good question! One could merge the values of multiple Victrons, but that would be... lame? It should not be too hard to just instanciate two instances of the Victron class handling the values.

@helgeerbe Have you tried using a SoftwareSerial for communicating with the Victron? I am planning on trying it, since the JK BMS needs 115200 baud, and that's not possible with SoftwareSerial. So HardwareSerials are becoming scarce.

**Reply by dirk-schulz** [reply]
that's a pity. Also not available for a potential beta tester? :-)
But many thanks for your work on this great project! then I will solve the connection to the 2nd victron it in the meanwhile via esphome and victron connection for HA.

**Reply by helgeerbe** [reply]
@schlimmchen I think that SoftwareSerial would be fine (Victron uses 19200 baud). Using SoftwareSerial would also simplify supporting multiple Victrons. 

Should be easy to instantiate multiple time the VictronFramehandler. At the end you would need only a simple array. This should also work with the Victron SmartShunt.

I think more work have to spend into the powerLimiter and the UserInterface. SmartShunt would be just another BMS.

**Reply by madmartin** [reply]
Yes folks, as I already wrote somewhere else I intend to expand OpenDTU-onbattery to support 2+ Victron MPPT loaders. Until then, it just works when you have SmartSolar MPPT loaders coupled together via Bluetooth (called VeSmart Networking) but the yield and MPPT power statistics shown is only for the connected  Victron loader.
Keep calm...

**Comment by madmartin** [comment]
another PCB layout is also in progress on my side...

**Comment by Manos1966** [comment]
> Yes folks, as I already wrote somewhere else I intend to expand OpenDTU-onbattery to support 2+ Victron MPPT loaders. Until then, it just works when you have SmartSolar MPPT loaders coupled together via Bluetooth (called VeSmart Networking) but the yield and MPPT power statistics shown is only for the connected Victron loader. Keep calm...

Hi @madmartin, sorry if I do not understand everything you wrote above,
does it mean that 
- one can have many Victron MPTTs connected together, via Bluetooth, they will work correctly to charge the battery.
- Victron yield and MPPT power statistics will be only the ones delivered only from the Victron that is connected to OpenDTU
- As result of the point above, Solar Pass Through will only use the information of the Victron  that is connected to OpenDTU 

Correct?
RGDS
Manos
